File: Source/NES/NESDSK.cpp
#include "NESDSK.h"
#include "NESEMU.h"

NESDSK::NESDSK(NESEMU *emu) : NESCMP(emu) {


}

NESDSK::~NESDSK() {
	// Destructor implementation
	if (prg_roms) {
		for (int i = 0; i < prg_rom_count; ++i) {
			delete[] prg_roms[i];
		}
		delete[] prg_roms;
	}
	if (chr_roms) {
		for (int i = 0; i < chr_rom_count; ++i) {
			delete[] chr_roms[i];
		}
		delete[] chr_roms;
	}
}

char* NESDSK::toJSON() {
	// Convert to JSON implementation
	char *json = new char[1024];
	SDL_snprintf(json, 1024, "{\"version\": %d, \"mapper\": %d, \"mirroring\": %d, \"tv_system\": %d, \"prg_rom_size\": %d, \"prg_rom_count\": %d, \"chr_rom_size\": %d, \"chr_rom_count\": %d}",
		version, mapper, mirroring, tv_system, prg_rom_size, prg_rom_count, chr_rom_size, chr_rom_count);

	return json;
}

void NESDSK::load(Uint8 *datas) {
	// Load ROM data into emulator
	if (datas[0] != 0x4E || datas[1] != 0x45 || datas[2] != 0x53 || datas[3] != 0x1A) {
		// Handle invalid ROM file
		SDL_Log("NESDSK: Invalid ROM file");
		return;
	}

	version = 1;
	if (datas[7] & 0x0C == 0x08) {
		version = 2;
	}
	mapper = (datas[6] >> 4) | (datas[7] & 0xF0);
	mirroring = datas[6] & 0x01;
	tv_system = datas[9] & 0x01;

	prg_rom_size = 16 * 1024; // 16 KB
	prg_rom_count = datas[4];
	prg_roms = new Uint8*[prg_rom_count];

	chr_rom_size = 8 * 1024; // 8 KB
	chr_rom_count = datas[5];
	chr_roms = new Uint8*[chr_rom_count];

	int addr = 0x0010; // Start of ROM data
	for (int i = 0; i < prg_rom_count; ++i) {
		prg_roms[i] = new Uint8[prg_rom_size];
		SDL_memcpy(prg_roms[i], &datas[addr], prg_rom_size);
		addr += prg_rom_size;
	}
	for (int i = 0; i < chr_rom_count; ++i) {
		chr_roms[i] = new Uint8[chr_rom_size];
		SDL_memcpy(chr_roms[i], &datas[addr], chr_rom_size);
		addr += chr_rom_size;
	}

	active_prg_rom[0] = 0; // Start with the first PRG ROM bank active
	active_prg_rom[1] = prg_rom_count > 1 ? 1 : 0; // If there's a second bank, use it; otherwise, mirror the first bank
	
	active_chr_rom[0] = 0; // Start with the first CHR ROM bank active
	active_chr_rom[1] = chr_rom_count > 1 ? 1 : 0; // If there's a second bank, use it; otherwise, mirror the first bank
}

File: Source/NESEMU/NESDSK.swift
import Foundation

enum NESDSKMirroring: Int {
    case vertical = 0
    case horizontal
}

class NESDSK: NESCOM, CustomStringConvertible {
    var mapper: Int = 0
    var mirroring: NESDSKMirroring = .vertical

    var prgRomCount = 0
    var prgRomSizeKB = 16
    var prgRoms: [[UInt8]] = []
    var prgRomActive = [ 0, 0]

    var chrRomCount = 0
    var chrRomSizeKB = 8
    var chrRoms: [[UInt8]] = []
    var chrRomActive = [ 0, 0]
    
    var description: String {
        return "NESDSK:\n\tMapper: \(mapper)\n\tMirroring: \(mirroring)\n\tPRGROM: \(prgRomCount) * \(prgRomSizeKB) KB\n\tCHRROM: \(chrRomCount) * \(chrRomSizeKB) KB"
    }
    
    func loadRom(_ data: [UInt8]) {
        if data[0] != 0x4E || data[1] != 0x45 || data[2] != 0x53 || data[3] != 0x1A {
            log("NESDSK - ERROR: Invalid iNES header")
            return
        }
        
        mapper = Int((data[7] & 0xF0) | ((data[6] & 0xF0) >> 4))
        mirroring = ((data[6] & 0x01) != 0) ? .vertical : .horizontal
        
        // Parse iNES header
        // Bytes 4: PRG ROM size in units of 16 KB
        prgRomCount = Int(data[4])
        // Byte 5: CHR ROM size in units of 8 KB
        chrRomCount = Int(data[5])
        
        // PRG ROM starts at offset 16
        var addr = 0x0010
        var size = prgRomSizeKB * 1024

        // Load PRG ROM
        for _ in 0..<prgRomCount {
            let prgRom = Array(data[addr..<(addr + size)])
            prgRoms.append(prgRom)
            addr += size
        }
        prgRomActive = [0, prgRomCount - 1]

        if chrRomCount == 0 {
            let chrRamBuffer = [UInt8](repeating: 0, count: 8 * 1024)
            chrRoms.append(chrRamBuffer)
            chrRomActive = [0, 0]
        } else {
            // Load CHR ROM
            size = chrRomSizeKB * 1024
            for _ in 0..<chrRomCount {
                let chrRom = Array(data[addr..<(addr + size)])
                chrRoms.append(chrRom)
                addr += size
            }
            chrRomActive = [0, 0]
        }
    }
    
    // Read from PRG ROM at offset (0x8000-FFFF, but we'll map the entire PRG ROM)
    func readPrgRom(_ addr: UInt16) -> UInt8 {
        let bankIndex = Int(addr & 0x3FFF)
        
        if addr < 0xC000 {
            return prgRoms[prgRomActive[0]][Int(bankIndex)]
        } else {
            return prgRoms[prgRomActive[1]][Int(bankIndex)]
        }
    }
    
    // Write to PRG ROM at offset (0x8000-FFFF, but we'll map the entire PRG ROM)
    func writePrgRom(_ addr: UInt16, _ value: UInt8) {
        if addr < 0xC000 {
            prgRoms[prgRomActive[0]][Int(addr & 0x3FFF)] = value
        } else {
            prgRoms[prgRomActive[1]][Int(addr & 0x3FFF)] = value
        }
    }
    
    func readChrRom(_ addr: UInt16) -> UInt8 {
        let bankIndex = Int(addr & 0x1FFF)
        return chrRoms[chrRomActive[0]][bankIndex]
    }
        
    func writeChrRom(_ addr: UInt16, _ value: UInt8) {
        let bankIndex = Int(addr & 0x1FFF)
        chrRoms[chrRomActive[0]][bankIndex] = value
    }
}
